File: ios/RNInBrowserWebViewManager.m
//
//  RNInBrowserWebViewManager.m
//  react-native-in-browser
//
//  Created by Le Ung on 29/10/25.
//

#import "RNInBrowserWebViewManager.h"
#import "RNInBrowserWebView.h"
#import <React/RCTUIManager.h>

@implementation RNInBrowserWebViewManager

RCT_EXPORT_MODULE(RNInBrowserWebView)

- (UIView *)view {
  return [[RNInBrowserWebView alloc] init];
}

RCT_EXPORT_VIEW_PROPERTY(url, NSString)
RCT_EXPORT_VIEW_PROPERTY(onLoadStart, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onLoadEnd, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onLoadError, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onUrlChange, RCTDirectEventBlock)

RCT_EXPORT_METHOD(goBack:(nonnull NSNumber *)reactTag) {
  [self.bridge.uiManager addUIBlock:^(RCTUIManager *uiManager, NSDictionary<NSNumber *, UIView *> *viewRegistry) {
    RNInBrowserWebView *view = (RNInBrowserWebView *)viewRegistry[reactTag];
    if ([view isKindOfClass:[RNInBrowserWebView class]]) {
      [view goBack];
    }
  }];
}

RCT_EXPORT_METHOD(goForward:(nonnull NSNumber *)reactTag) {
  [self.bridge.uiManager addUIBlock:^(RCTUIManager *uiManager, NSDictionary<NSNumber *, UIView *> *viewRegistry) {
    RNInBrowserWebView *view = (RNInBrowserWebView *)viewRegistry[reactTag];
    if ([view isKindOfClass:[RNInBrowserWebView class]]) {
      [view goForward];
    }
  }];
}

RCT_EXPORT_METHOD(reload:(nonnull NSNumber *)reactTag) {
  [self.bridge.uiManager addUIBlock:^(RCTUIManager *uiManager, NSDictionary<NSNumber *, UIView *> *viewRegistry) {
    RNInBrowserWebView *view = (RNInBrowserWebView *)viewRegistry[reactTag];
    if ([view isKindOfClass:[RNInBrowserWebView class]]) {
      [view reload];
    }
  }];
}

RCT_EXPORT_METHOD(stopLoading:(nonnull NSNumber *)reactTag) {
  [self.bridge.uiManager addUIBlock:^(RCTUIManager *uiManager, NSDictionary<NSNumber *, UIView *> *viewRegistry) {
    RNInBrowserWebView *view = (RNInBrowserWebView *)viewRegistry[reactTag];
    if ([view isKindOfClass:[RNInBrowserWebView class]]) {
      [view stopLoading];
    }
  }];
}

@end

File: android/src/main/java/com/rninbrowser/RNInBrowserWebViewManager.java
package com.rninbrowser;

import android.view.ViewGroup;
import android.view.ViewGroup.LayoutParams;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.common.MapBuilder;
import com.facebook.react.uimanager.SimpleViewManager;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.annotations.ReactProp;

import java.util.Map;

public class RNInBrowserWebViewManager extends SimpleViewManager<RNInBrowserWebView> {

  public static final String REACT_CLASS = "RNInBrowserWebView";

  @NonNull
  @Override
  public String getName() {
    return REACT_CLASS;
  }

  @NonNull
  @Override
  protected RNInBrowserWebView createViewInstance(@NonNull ThemedReactContext reactContext) {
    RNInBrowserWebView webView = new RNInBrowserWebView(reactContext);

    // ✅ Cho phép WebView match với style từ React Native (flex hoặc width/height)
    ViewGroup.LayoutParams params = new ViewGroup.LayoutParams(
        LayoutParams.MATCH_PARENT,
        LayoutParams.MATCH_PARENT
    );
    webView.setLayoutParams(params);

    return webView;
  }

  @ReactProp(name = "url")
  public void setUrl(RNInBrowserWebView view, @Nullable String url) {
    if (url != null && !url.isEmpty()) {
      view.setUrl(url);
    }
  }

  @Nullable
  @Override
  public Map<String, Object> getExportedCustomDirectEventTypeConstants() {
    return MapBuilder.<String, Object>builder()
        .put("onLoadStart", MapBuilder.of("registrationName", "onLoadStart"))
        .put("onLoadEnd", MapBuilder.of("registrationName", "onLoadEnd"))
        .put("onLoadError", MapBuilder.of("registrationName", "onLoadError"))
        .put("onUrlChange", MapBuilder.of("registrationName", "onUrlChange"))
        .build();
  }

  @Override
  public Map<String, Integer> getCommandsMap() {
    return MapBuilder.of(
        "goBack", 1,
        "goForward", 2,
        "reload", 3,
        "stopLoading", 4
    );
  }

  @Override
  public void receiveCommand(@NonNull RNInBrowserWebView view, int commandId, @Nullable ReadableArray args) {
    switch (commandId) {
      case 1: // goBack
        if (view.canGoBack()) view.goBack();
        break;
      case 2: // goForward
        if (view.canGoForward()) view.goForward();
        break;
      case 3: // reload
        view.reload();
        break;
      case 4: // stopLoading
        view.stopLoading();
        break;
    }
  }
}
